libtransmission/verify.cc

```diff
@@ -83,23 +102,59 @@ void tr_verify_worker::verify_torrent(Mediator& verify_mediator, std::atomic<boo
         piece_pos += bytes_this_pass;
         file_pos += bytes_this_pass;
 
+        if (left_in_file == 0)
+        {
+            /* We want all on_file_error() calls for a file index to be consecutive. Therefore
+             * don't add files for which an error has already been emitted. We don't need to report
+             * piece errors for missing or wrong length files.
+             */
+            if (file_valid)
+            {
+                files_ending_in_current_piece.push_back(file_index);
+            }
+        }
+
         /* if we're finishing a piece... */
         if (left_in_piece == 0)
         {
             auto const has_piece = sha->finish() == metainfo.piece_hash(piece);
             verify_mediator.on_piece_checked(piece, has_piece);
+            if (has_piece)
+            {
+                for (auto const file : files_ending_in_current_piece)
+                {
+                    verify_mediator.on_file_ok(file);
+                }
+            }
+            else
+            {
+                auto const error{ fmt::format("Piece {:d} hash fail", piece) };
+                for (auto const file : files_ending_in_current_piece)
+                {
+                    verify_mediator.on_file_error(file, error);
+                }
+                if (left_in_file != 0)
+                {
+                    verify_mediator.on_file_error(file_index, error);
+                }
+                file_valid = false;
+            }
 
-            /* sleeping even just a few msec per second goes a long
-             * way towards reducing IO load... */
-            if (auto const now = current_time_secs(); last_slept_at != now)
+            if (sleep_ms_per_second > 0ms)
             {
-                last_slept_at = now;
-                std::this_thread::sleep_for(SleepPerSecondDuringVerify);
+                /* sleeping even just a few msec per second goes a long
+                 * way towards reducing IO load... */
+                if (auto const now = current_time_secs(); last_slept_at != now)
+                {
+                    last_slept_at = now;
+                    std::this_thread::sleep_for(SleepPerSecondDuringVerify);
+                }
```

Comment:
Maybe you can enlighten me about the reasonings for allowing to disable the 100ms breaks.
Judging from this comment, it sounds to me sleeping every once in a while is always a good idea:

https://github.com/transmission/transmission/blob/0f321b850342646c8ca3fd5e54a75220e34865cf/libtransmission/verify.cc#L145-L146

If we all agree in the end that allowing to disable the 100ms breaks is a good idea, then `sleep_ms_per_second` should either be replaced with a `bool` parameter, or be passed as a parameter to `sleep_for()`. The current code here looks weird.

Reply:
My understanding of that comment is that when running a session with other torrents downloading, you want to ensure the verify does not slow down the other torrents. The change I am adding is for a command line utility that runs in a single thread with no other IO. In this case the user expects the verify to run as quick as possible, similar to running other IO intensive utils like `cp` or `tar`.

I intended to pass `sleep_ms_per_second` to `sleep_for()`. That was a mistake.

Reply:
> In this case the user expects the verify to run as quick as possible

True, but as a user, I'd appreciate it if Transmission does not starve my other IO-intensive applications, or cause an overall slowdown in any way. So I think it's better to revert this change.

Reply:
You can use `ionice` if you want to limit IO. I would prefer not to slow it down when I just want it to run as fast as possible. Could be add a command line option to limit the IO?

Reply:
Thoughts Charles?

Reply:
Hmm, looks like Linux can use [ioprio_set](https://man7.org/linux/man-pages/man2/ioprio_set.2.html) and Windows can use [SetProcessInformation](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessinformation).

Reply:
> Thoughts Charles?

IMO our `std::this_thread::sleep_for()` call is a little crude, but it's a one-liner that is working fine, is portable code, needs no error-handling and no ifdefs. I'd keep it as-is unless there's a known issue with it.

> My understanding of that comment is that when running a session with other torrents downloading, you want to ensure the verify does not slow down the other torrents.

`tr_verify_worker` runs in its own thread, so there shouldn't be any issue of verify blocking the session thread.